### ISSUE TITLE
🐛 BUG: Using both `--route` and `--remote` causes WebSocket error

### ISSUE THREAD

**Issue by helloimalastair** [cloudflare/workers-sdk#3755]
### Which Cloudflare product(s) does this pertain to?

Wrangler core

### What version(s) of the tool(s) are you using?

3.5.0

### What version of Node are you using?

20.5.1

### What operating system are you using?

macOS Ventura 13.5

### Describe the Bug

1. Create `index.js` file:
```js
export default {
	fetch(req) {
		const url = new URL(req.url);
		return new Response(`Your hostname is: ${url.hostname}`);
	}
}
```
2. Run Wrangler Dev with both `--remote` and `--route` flags(with a zone you have access to):
```sh
$ wrangler dev index.js --route "goalastair.com/*" --remote
 ⛅️ wrangler 3.5.0
------------------
⬣ Listening at http://0.0.0.0:8787
- http://127.0.0.1:8787
- http://192.168.32.2:8787
Total Upload: 5.96 KiB / gzip: 1.84 KiB
▲ [WARNING] worker failed to prewarm:  Bad Gateway


Waiting for connection...
╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ [b] open a browser, [d] open Devtools, [l] turn on local mode, [c] clear console, [x] to exit                                                                                                                  │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:30632
            throw a;
            ^

Error: WebSocket was closed before the connection was established
    at WebSocket2.close (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:108521:11)
    at close (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:127122:14)
    at /Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:127230:7
    at jh (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:29688:15)
    at exports2.unstable_runWithPriority (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:25660:16)
    at Pc (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:26308:16)
    at Qg (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:29651:18)
    at /Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:29640:13
    at V (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:25585:22)
    at Timeout.w [as _onTimeout] (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:25420:13)
Emitted 'error' event on WebSocket instance at:
    at emitErrorAndClose (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:108968:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)

Node.js v20.5.1
```

**Comment by iksaku** [comment]
Been waiting on this issue for some months now as well. My workaround was to comment out `[route]` definition from `wrangler.toml` each time I was running my projects locally, but also needed to remember to un-comment before any commits/deploys...

Just discovered today that you can pass `--routes=[]` on your command via `package.json` scripts and it will overwrite the definition from `wrangler.toml`.

So, the config may look like this:

`wrangler.toml`:
```toml
name = "my-worker"
main = "src/worker.ts"
compatibility_date = "2023-10-23"

workers_dev = false

[route]
  pattern = "worker.example.com"
  custom_domain = true
```

`package.json`:
```json
{
  "scripts": {
    "dev": "wrangler dev --remote --routes=[]"
  }
}
```

Hopefully we can get come kind fix, or at least, a flag to bypass `routes` when using `--remote`.

**Comment by penalosa** [comment]
@helloimalastair @iksaku I'm unable to reproduce this in either the latest Wrangler or Wrangler 3.5. A couple questions to help debug this:
* Are you using Cloudflare Access?
* Is the route you're trying currently orange-clouded on the Cloudflare dashboard?
* Are you logged in as a user who has access to that zone? (i.e. what does `wrangler whoami` give you?)

**Comment by helloimalastair** [comment]
@penalosa
1. Using Access? No
2. Orange-Clouded? Yes
3. Logged in? Yes
Just did some digging, and it looks like `goalastair.com` is subject to a Bulk Redirect rule, and if I switch to `www.goalastair.com`, it does work.
Also found a separate bug, in that when hitting the `b` key to open in browser, it spawns `www.goalastair.com:8787`

**Comment by penalosa** [comment]
Thanks @helloimalastair —in that case I'll close this issue. The hotkey bug is captured in https://github.com/cloudflare/workers-sdk/issues/3907, I think

**Comment by helloimalastair** [comment]
@penalosa So wait, is it expected that `--remote` doesn't work when rules are applied to the target? Or should that be documented somewhere?

**Comment by penalosa** [comment]
`--remote` should follow the same logic as deployed workers, as far as I know. Because redirect rules run before Workers, if you have:
* a redirect setup from `source.example.com` to `target.example.com`
* a worker with a route of `source.example.com`
Accessing `source.example.com` won't execute your worker. You'd have to give the worker a route of `target.example.com` (the target of the redirect rule) for the worker to execute.

**Comment by lucas-pelton** [comment]
The OP may have a redirect rule set up that's not a best practice and is leading to this error...

I have a similar error, but without the redirect rules. My setup is this

- worker1 @ domain.com/*
- worker2 @ api.domain.com/path/*
- worker @ telcom.domain.com/path/*

If I have any routes in my wrangler.toml is when the OP's bug surfaces. If I add routes via the dashboard and comment out the routes in wrangler.toml, I'm able to get wrangler dev --remote running (or wrangler dev, press l to turn off local)

Interestingly, the error message is different if I use the --remote flag vs switching to remote. Switching to remote actually throws an error that was mentioned in a bug closed for Wrangler v2. I can provide more details here or elsewhere if this piques anyone's interest.

**Comment by penalosa** [comment]
@lucas-pelton Would you be able to file a new issue with more details, and the exact error messages you're receiving? That sounds like a bug we should fix

**Comment by lucas-pelton** [comment]
Let me put together a proper test and report using a domain that's not running live services. It may be a minute, but I'll do it.

**Comment by iksaku** [comment]
Hello @penalosa,

1. Not using CF Access.
2. DNS entry is indeed orange-clouded as it shows is being proxied to my worker.
3. Yes, I'm logged in to the account owning the zone and worker.

In my case, I have no bulk redirects nor page rules configured for the zone, so a redirection misconfiguration should not be the problem here.
Also, I'm using a subdomain for my worker, not the root domain, and the `wrangler.toml` looks like this:

```toml
name = "my-worker"
main = "src/worker.ts"
compatibility_date = "2023-10-23"

workers_dev = false

[[r2_buckets]]
  binding = "BUCKET"
  bucket_name = "..."
  preview_bucket_name = "..."

[route]
  pattern = "worker.<my-domain>.com"
  custom_domain = true
```

Any ideas?

**Comment by lucas-pelton** [comment]
@iksaku 
The Wrangler docs suggest you can't have these in the same config:
```
workers_dev = false

[route]
  pattern = "worker.<my-domain>.com"
  custom_domain = true
```
Specifically

> If you change your routes in the dashboard, Wrangler will override them in the next deploy with the routes you have set in your wrangler.toml. To manage routes via the Cloudflare dashboard only, remove any route and routes keys from your wrangler.toml file. Then add workers_dev = false to your wrangler.toml file.
https://developers.cloudflare.com/workers/wrangler/configuration/

Is this related to your use of `wrangler dev --routes=[]` workaround?

**Comment by iksaku** [comment]
I must've skipped that in the docs.

And yes, this is related to my workaround

**Comment by iksaku** [comment]
@penalosa Sure!
My account ID is `df4989e4945becf8236d033213f17af8`.

<details>
<summary>Attached error log</summary>

```
> wrangler dev --remote

 ⛅️ wrangler 3.14.0
-------------------
Your worker has access to the following bindings:
- R2 Buckets:
  - BUCKET: ...
⬣ Listening at http://0.0.0.0:8787
- http://127.0.0.1:8787
- http://192.168.1.81:8787
- http://100.70.222.84:8787
- http://198.19.249.3:8787
Total Upload: 48.00 KiB / gzip: 11.86 KiB
Waiting for connection...

Error: WebSocket was closed before the connection was established
    at WebSocket2.close (./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:107657:18)
    at close (./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:127819:14)
    at ./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:127880:7
    at jh (./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:28430:15)
    at exports2.unstable_runWithPriority (./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:24402:16)
    at Pc (./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:25050:16)
    at Qg (./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:28393:18)
    at ./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:28382:13
    at V (./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:24327:22)
    at Timeout.w [as _onTimeout] (./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:24162:13)
Emitted 'error' event on WebSocket instance at:
    at emitErrorAndClose (./node_modules/.pnpm/wrangler@3.14.0/node_modules/wrangler/wrangler-dist/cli.js:108144:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
```
</details>

If I instead run using my workaround (`wrangler dev --remote --routes=[]`) or directly comment out the `[route]` definition in my `wrangler.toml` file, the error in question disappears and the worker keeps working as intended.

Thank you!

**Comment by lucas-pelton** [comment]
@iksaku I get a very similar error log running `wrangler dev --remote`, but if I run `wrangler dev` and then press `l` to switch from local to remote, the error is significantly different.

This JSON parsing error is the one that I mentioned in an earlier post that is similar to the Wrangler v2 bug at #2581 (at least in that they're trying to JSON.parse an HTML file when trying launch a remote dev server.

Stack trace:
```
✘ [ERROR] Error while creating remote dev session: ParseError: Unexpected token <

      at parseJSON
  (/home/runcloud/wrangler-projects/telcom-processor/node_modules/wrangler/wrangler-dist/cli.js:119580:11)
      at createPreviewSession
  (/home/runcloud/wrangler-projects/telcom-processor/node_modules/wrangler/wrangler-dist/cli.js:150592:51)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async start
  (/home/runcloud/wrangler-projects/telcom-processor/node_modules/wrangler/wrangler-dist/cli.js:150781:9)
```

My HTML file is a cloudflare error page - Error 1031: Improperly configured Workers Preview for dev-telcom.eaevsf.us. You may need to reinitialize.

Interestingly, I've commented out routes in wrangler.toml and I'm getting this error when _switching to_ remote, but not when using the `--remote` flag. 

This is occurring on the sub.domain/* worker while I have a domain.com/* worker manually configured on the dashboard. (This is the live services setup, so I can't test the effect of taking down the domain.com/* route binding.

**Comment by flexchar** [comment]
This is extremely frustrating as workerd is still encountering some issues that are not present when deployed. I'd really appreciate this being looked on more. 

I often get:
`⚠️ [WARNING] worker failed to prewarm:  Bad Gateway`
OR
`✘ [ERROR] Error while creating remote dev session: ParseError: Unexpected token <

      at parseJSON
`

The workaround using `npx wrangler dev --remote --routes "[]"` does help :)

**Comment by penalosa** [comment]
@lucas-pelton @iksaku @flexchar We've recently added some more error reporting into Wrangler, which should significantly help with debugging this. Would one of you be able to upload the error log that Wrangler outputs at the end of a run?

**Comment by flexchar** [comment]
Hey Somhairle, I launched as typical and pressed `l` to disable local mode receiving exact same error. I'm on the `3.25.0` version. Where should I look for the new logs?

**Comment by penalosa** [comment]
@lucas-pelton this should be fixed in #5107
